### ISSUE TITLE
Undo functionality in the editor

### ISSUE THREAD

**Issue by tisto** [plone/volto#1954]
Editors always ask us for the option to undo a change in the edit mode. I understand this is non-trivial to implement but I wanted to add it here anyways.

**Comment by tiberiuichim** [comment]
I've been looking at this. I see the following roadmap:

- first cleanup the Form.jsx implementation. Split the Visual form to a separate component, split the utility methods to a new helper. This work has been done in [volto-blocks-form](https://github.com/eea/volto-blocks-form) (but probably needs to be refreshed).
- decide what type of undo we want. Some blocks, such as Text block have their own built-in undo that work at "block level". 

We could implement undo at block level for the other blocks and have the "whole-form undo" treat blocks atomically and understand atomicity based on the block identity (which block was changed last?)

The other method would imply that we keep snapshots of the whole form state and undo using them.

I'm worries about memory performance. A naive approach could balloon RAM usage very fast. I've looked at immer, it can represent state changes as patches, which could help. The way to go, in this case, would be: refactor the Form.jsx to simplify the whole thing, refactor the state storage to use immer, implement an UndoManager that keeps the intermediary states.

Another approach would be to abandon the form state and use the redux store using the approach in the [official docs](https://redux.js.org/recipes/implementing-undo-history). This would require a bigger refactoring, as far as I see.
